### PR TITLE
TST: bare pytest raises in tests/scalar

### PR DESCRIPTION
**Pull request by quangngd**
- [ ] ref #30999
- [ ] tests added / passed
- [ ] passes `black pandas`
- [ ] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [ ] whatsnew entry

https://github.com/pandas-dev/pandas/issues/30999

### CODE REVIEW

File: pandas/tests/scalar/period/test_period.py [hunk 1]
```diff
@@ -1030,7 +1038,8 @@ def test_sub_delta(self):
         result = left - right
         assert result == 4 * right.freq
 
-        with pytest.raises(IncompatibleFrequency):
+        msg = "Input has different freq=M from Period\\(freq=A-DEC\\)"
```
Comment (MomIsBestFriend) [inline pandas/tests/scalar/period/test_period.py]: Nitpick, not a blocker.

```suggestion
        msg = r"Input has different freq=M from Period\(freq=A-DEC\)"
```

File: pandas/tests/scalar/test_na_scalar.py [hunk 2]
```diff
@@ -145,7 +147,7 @@ def test_logical_and():
     assert False & NA is False
     assert NA & NA is NA
 
-    with pytest.raises(TypeError):
+    with pytest.raises(TypeError, match="unsupported operand type"):
```
Comment (MomIsBestFriend) [inline pandas/tests/scalar/test_na_scalar.py]: Can you take the value of ```match``` outside the ```pytest.raises```

```suggestion
    msg = "unsupported operand type"
    with pytest.raises(TypeError, match=msg):
```

File: pandas/tests/scalar/interval/test_interval.py [hunk 3]
```diff
@@ -255,5 +255,9 @@ def test_constructor_errors_tz(self, tz_left, tz_right):
         left = Timestamp("2017-01-01", tz=tz_left)
         right = Timestamp("2017-01-02", tz=tz_right)
         error = TypeError if com.any_none(tz_left, tz_right) else ValueError
-        with pytest.raises(error):
+        msg = (
+            "left and right must have the same time zone|"
```
Comment (simonjayhawkins) [inline pandas/tests/scalar/interval/test_interval.py]: rather than use the regex, could expand the if com.any_none instead

**Review by ShaharNaveh — CHANGES_REQUESTED** [review]
Thank you so much for taking some of these!

I have some minor comments

File: pandas/tests/scalar/period/test_period.py [hunk 4]
```diff
@@ -1222,14 +1244,20 @@ def test_add_offset(self):
...
                 timedelta(hours=23),
             ]:
-                with pytest.raises(IncompatibleFrequency):
+                msg = "Input has different freq|Input cannot be converted to Period"
+                with pytest.raises(IncompatibleFrequency, match=msg):
                     p + o
 
-                if isinstance(o, np.timedelta64):
-                    with pytest.raises(TypeError):
+                if isinstance(o, np.timedelta64,):
```
Comment (simonjayhawkins) [inline pandas/tests/scalar/period/test_period.py]: was this change intentional?

File: pandas/tests/scalar/timestamp/test_constructors.py [hunk 5]
```diff
@@ -225,15 +231,20 @@ def test_constructor_positional(self):
 
     def test_constructor_keyword(self):
         # GH 10758
-        with pytest.raises(TypeError):
+        msg = "function missing required argument 'day'|Required argument 'day'"
```
Comment (simonjayhawkins) [inline pandas/tests/scalar/timestamp/test_constructors.py]: do you know where these inconsistencies arise?

File: pandas/tests/scalar/timestamp/test_constructors.py [hunk 6]
```diff
@@ -353,25 +367,28 @@ def test_out_of_bounds_value(self):
...
-        with pytest.raises(ValueError):
+        with pytest.raises(ValueError, match=msg):
             Timestamp("2263-01-01")
 
     def test_barely_out_of_bounds(self):
         # GH#19529
         # GH#19382 close enough to bounds that dropping nanos would result
         # in an in-bounds datetime
-        with pytest.raises(OutOfBoundsDatetime):
+        msg = "Out of bounds"
```
Comment (simonjayhawkins) [inline pandas/tests/scalar/timestamp/test_constructors.py]: is there more to this message?

File: pandas/tests/scalar/timestamp/test_constructors.py [hunk 7]
```diff
@@ -382,7 +399,8 @@ def test_bounds_with_different_units(self):
         for date_string in out_of_bounds_dates:
             for unit in time_units:
                 dt64 = np.datetime64(date_string, unit)
-                with pytest.raises(ValueError):
+                msg = "Out of bounds"
+                with pytest.raises(ValueError, match=msg):
```
Comment (simonjayhawkins) [inline pandas/tests/scalar/timestamp/test_constructors.py]: is this OutOfBoundsDatetime

File: pandas/tests/scalar/timestamp/test_timezones.py [hunk 8]
```diff
@@ -26,14 +26,14 @@ def test_tz_localize_pushes_out_of_bounds(self):
         pac = Timestamp.min.tz_localize("US/Pacific")
         assert pac.value > Timestamp.min.value
         pac.tz_convert("Asia/Tokyo")  # tz_convert doesn't change value
-        with pytest.raises(OutOfBoundsDatetime):
+        with pytest.raises(OutOfBoundsDatetime, match="^$"):
```
Comment (simonjayhawkins) [inline pandas/tests/scalar/timestamp/test_timezones.py]: would it make sense to have a message here?

File: pandas/tests/scalar/timestamp/test_timezones.py [hunk 9]
```diff
@@ -327,14 +339,16 @@ def test_timestamp_constructor_near_dst_boundary(self):
...
+            with pytest.raises(pytz.AmbiguousTimeError, match=msg):
                 Timestamp("2015-10-25 02:00", tz=tz)
 
         result = Timestamp("2017-03-26 01:00", tz="Europe/Paris")
         expected = Timestamp("2017-03-26 01:00").tz_localize("Europe/Paris")
         assert result == expected
 
-        with pytest.raises(pytz.NonExistentTimeError):
+        msg = "2017-03-26 02:00"
+        with pytest.raises(pytz.NonExistentTimeError, match=msg):
```
Comment (simonjayhawkins) [inline pandas/tests/scalar/timestamp/test_timezones.py]: ok to inline short, single use messages

**Review by simonjayhawkins — COMMENTED** [review]
Thanks @quangngd for the PR. generally looks good on first pass. in future can you limit the changes to batches of no more than 5 files.

File: pandas/tests/scalar/test_na_scalar.py (hunk 2 above)
Comment (simonjayhawkins) [reply pandas/tests/scalar/test_na_scalar.py]: @MomIsBestFriend I might have missed a discussion on this, but we use this pattern a lot for short, single use messages.

Comment (MomIsBestFriend) [reply pandas/tests/scalar/test_na_scalar.py]: Oh didn't know about that, thank you for informing me @simonjayhawkins

File: pandas/tests/scalar/period/test_period.py (hunk 4 above)
Comment (quangngd) [reply pandas/tests/scalar/period/test_period.py]: sorry, this was unintentional

File: pandas/tests/scalar/test_na_scalar.py (hunk 2 above)
Comment (simonjayhawkins) [reply pandas/tests/scalar/test_na_scalar.py]: no problem, if we don't have a standard, is a matter of personal preference. we started adding the messages before we adopted black, so the norm was to have a separate assignment to improve the formatting. I would not be adverse to in-lining for all single use messages now as black would keep the longer messages tidy.

File: pandas/tests/scalar/timestamp/test_timezones.py (hunk 8 above)
Comment (quangngd) [reply pandas/tests/scalar/timestamp/test_timezones.py]: the current implementation returns nothing, it's indeed should better have some message but that's out of the issue's scope

File: pandas/tests/scalar/timestamp/test_constructors.py (hunk 7 above)
Comment (quangngd) [reply pandas/tests/scalar/timestamp/test_constructors.py]: yes it is, sorry i just didn't pay attention to the old code

File: pandas/tests/scalar/timestamp/test_constructors.py (hunk 6 above)
Comment (quangngd) [reply pandas/tests/scalar/timestamp/test_constructors.py]: the full message is "Out of bounds nanosecond timestamp: 2262-04-11 23:47:16".
I will add the full message if needed to. Looking at the old code from other places I thought this is enough.

File: pandas/tests/scalar/timestamp/test_constructors.py (hunk 5 above)
Comment (quangngd) [reply pandas/tests/scalar/timestamp/test_constructors.py]: if we only have ```function missing required argument 'day'```
it would only pass the Linux py36_locale_slow_old_np and ```Linux py37_locale``` azure pipeline
The error was raised by pandas/_libs/tslibs/timestamps.pyx:467
```
ts_input = datetime(**datetime_kwargs)
```
Maybe it's a python thing?

Comment (quangngd) [reply pandas/tests/scalar/timestamp/test_constructors.py]: I know all of our pipeline run on py 3 but when looking around, i noticed that when run
```
open()
```
Python 3 (3.7.3) would returns ```TypeError: open() missing required argument 'file' (pos 1)``` and
Python 2 (2.7.17) would returns ```TypeError: Required argument 'name' (pos 1) not found```

Comment (simonjayhawkins) [reply pandas/tests/scalar/timestamp/test_constructors.py]: ok, this is an indication that the message is not raised by pandas. what's the policy here @MomIsBestFriend?

File: pandas/tests/scalar/timestamp/test_timezones.py (hunk 8 above)
Comment (simonjayhawkins) [reply pandas/tests/scalar/timestamp/test_timezones.py]: yep, out of scope. but if you could raise an issue, that'll be great.

**Review by simonjayhawkins — APPROVED** [review]
Thanks @quangngd lgtm. @MomIsBestFriend ok to merge?

**Comment by simonjayhawkins** [comment]
Thanks @quangngd
